### DISCUSSION TITLE
tvp for MHE and MPC

### DISCUSSION THREAD

**Discussion by SR-dTdt** [do-mpc/do-mpc#151] [Q&A] [answered]
Hello,

1. The .prediction command provides the possibility to access _tvps of the controller. Is there a similar functionality to access these for the moving horizon estimator?
2. Let's assume that n is the horizon for MHE and MPC. MPC requires n+1 tvps whereas MHE requires n tvps. Is this occurence related to the meyer term which is part of the MPC objective function?
3. Taking x_0 as a reference point. My understanding is, that we need to provide tvps up to x_(-1) for mhe? Measurements within the y_fun function need to be provided up to x_0?

Best,
SR

**Comment by 4flixt** [comment, marked as answer]
Hi @SR-dTdt,

you probably refer to this [function](https://www.do-mpc.com/en/latest/api/do_mpc.data.MPCData.prediction.html)? 
It is indeed only available for the MPC controller atm. 
Of course the sequence of tvps used at each iteration of the MHE also exist but there is no public API atm. I will add an issue to make them available.

Having N+1 time-varying parameters is a more recent change of the MPC controller. It is mostly used for the meyer term in the cost function as you said (e.g. when you do set-point tracking you also want to penalize the difference of the finale state (N+1) to the setpoint (the tvp).

If you do estimation with the MHE it helps to have [this figure ](https://www.do-mpc.com/en/latest/theory_mhe.html#moving-horizon-estimation-problem) in mind. From a sequence of N+1 states we only obtain N measurements as generally the measurement function is a function of x (N+1 elements) and u (N elements). For the same reason I believe it makes sense to only have N time-varying parameters from x_0 to N. 

I hope this answers your question!

Best
4flixt

**Reply by 4flixt** [reply]
Voila the new issue: #152 .

**Reply by SR-dTdt** [reply]
Hi, 
thx a lot. 

Having the figure you suggested in mind: 

1. MHE x_0 refers to the state at t_0?

2.MPC x_0 refers to the state at t_now?

Best,
SR

**Reply by 4flixt** [reply]
That is exactly right! In both, MHE and MPC, time is running in positive direction. The index k+1 is always at a later time then the index k.

**Reply by SR-dTdt** [reply]
This provides a much clearer picture. Thanks a lot :)
